### PR TITLE
Add Contextual interface, onScheduleContextualHook

### LINKED ISSUE #2980: Introduce a way to optionally pass `ContextView` to `Scheduler`s
The `Scheduler` interface has 3 submit methods and 3 mirroring methods on the `Scheduler.Worker` interface.
Additionally, all *vanilla* `Scheduler` used in Reactor come from the `Schedulers` factory, which offers a hooking mechanism.

This hooking mechanism is used by our own implementations, of course, but is also exposed for custom implementations to use. This cannot be enforced but is strongly recommended.

Unfortunately, the hooking signature is currently `onScheduleHook(String key, Function<Runnable, Runnable> hook)`. A `Scheduler` has no notion of `ContextView` because it can technically be used outside of operators without any contextual data associated to the submitted tasks.

This issue explores possible improvements to support submitting tasks with an associated `ContextView`.
It is up to operator implementations that submit tasks to a `Scheduler` to ensure in some way that their `currentContext()` is passed alongside the task.

This is the main challenge / design point.

# Introduce new type of `Schedulers` hook

First experimentation shows that it would be possible to avoid new APIs at `Scheduler` level by introducing the notion of a `ContextualRunnable` and contextual hooks.

Operators that use schedulers should ensure the submitted tasks are not simple instances of `Runnable` but rather of `ContextualRunnable`, adding a `contextView()` getter to the runnable.

Methods in the `Schedulers` class like `directSchedule` are a package-private common implementation for most vanilla `Scheduler`s that already perform some task wrapping and apply the basic hook(s). These could detect a `ContextualRunnable` and instead call a new category of hooks:

A variant of the hook that uses a `BiFunction<Runnable, ContextView, Runnable>` to expose a `ContextView` to the hook(s) _when it makes sense_. The `Context.empty()` singleton can be used to denote the task doesn't have any notion of context (in which case some hooks _may_ choose to shortcircuit and return the original `Runnable`).

Most of the time, the operator's subscriber implementation itself is the `Runnable`, so this simply translates to having a `contextView()` method implementation that returns the `currentContext()`.

In other cases, where a `Runnable` is built by the operator, the operator will need to capture its `Context` to create an instance of `ContextualRunnable` to which it can attach it.

# Alternative Options

 - add 3 overloads to `Scheduler` and 3 to `Worker`
 - operators have to explicitly pass in the `ContextView` alongside the `Runnable`

Challenges:
 - we need to also double the 6 `directSchedule` and `workerSchedule` supporting methods in `Schedulers`
 - 🎨 we need an acceptable `default` implementation
 - 🔨  we need to figure an acceptable mechanism for 3rd parties to use the `ContextView` to most likely decorate tasks
  - ⚠️ `Schedulers.Factory` currently doesn't allow reuse of original `Factory` nor does it allow composite factories
 

Although scheduling operators generally accept a `Scheduler` as input, by default they rely on static factory methods in `Schedulers`.

The methods delegate to one global `Factory` that can be replaced (`Schedulers.setFactory`). It is possible to restore the previous `Factory` indirectly by use of a `Schedulers.Snapshot` (as returned by `Schedulers.setFactoryWithSnapshot(Factory)`) and `Schedulers.resetFrom(Snapshot)`.

This allows **complete replacement** of original `Schedulers` implementations (see `VirtualTimeScheduler`), but **not composition**: one cannot create a `Scheduler` implementation that wraps the vanilla ones.

In order to keep the vanilla implementation and augment it, schedulers hook are offered, also stored statically in `Schedulers`.
Vanilla implementations will directly apply these hooks, whether or not they are the instances currently returned by the factory.

(note this can only happen in a corner case where a `Scheduler` instance is captured in a local variable, then a new implementation is injected via `Schedulers.setFactoryWithSnapshot` which doesn't shutdown the `Scheduler`. at this point, using the local variable still applies the hook(s)). 

The previous alternative would rely on changes at the `Schedulers` level in order to inject a `Factory` that _decorates_ the original one.

Several limitations and challenges arise:
 - no access to the "vanilla `Factory`"
 - maintain internal `CachedScheduler` (used to negate `dispose()` calls on common schedulers)
 - coupling with legacy `onScheduleHook`
 - do we need more than one level of decoration?
 - how to remove the decoration? if more than one level, how to remove it _selectively_ ?
 - `Schedulers` has two roles:
   - holding default schedulers (`parallel()`, `single()`...)
   - create detached instances (including via convenience factory methods) ultimately delegating to the currently set `Factory` (`newParallel(String, int)`, `newSingle(String)`...)
 - no interface to represent the "holding" role above and thus no way to reuse them. closest thing is `Snapshot` which doesn't publicly expose the schedulers but allows to restore them

### PR DESCRIPTION
**Pull request by simonbasle**
This PR adds the `Contextual` interface to denote an object that has a
`ContextView` attached.

It also adds a new variant of the `Schedulers#onScheduleHook` that takes a
`BiFunction<Runnable, ContextView, Runnable>`, allowing to craft hooks that can
receive the context associated with a given task

The choice of `BiFunction` rather than letting the hook check if the input
`Runnable` is also `Contextual` is made so that the `ContextView` is always in
scope of hooks, even if the individual named hooks don't maintain the Contextual
nature of the original `Runnable` task.

Note that individual operators which schedule a `Runnable` (be it themselves or
a lambda) will need to ensure that said `Runnable` is also a `Contextual` in
order for the hook to do its job.

Fixes #2980.

As a start in this PR, `publishOn` is migrated to use the `Contextual` trait.

TODO:
 - [x] have all classes which have a `contextView()` method implement `Contextual`
 - [x] migrate all operator subscribers that schedule themselves thanks to implementing `Runnable` to also implement `Contextual`
 - [x] ensure all operator subscribers scheduling a separate `Runnable` use a `Contextual` variant which either captures their current context or better yet delegate to their own `currentContext()` in a lazy fashion (eg. via method reference)

### CODE REVIEW

**Comment by simonbasle** [comment]
cc @marcingrzejszczak

File: gradle/libs.versions.toml [hunk 1]
```diff
@@ -20,6 +20,7 @@ blockhound = "io.projectreactor.tools:blockhound:1.0.6.RELEASE"
 byteBuddy-agent = { module = "net.bytebuddy:byte-buddy-agent", version.ref = "bytebuddy" }
 byteBuddy-api = { module = "net.bytebuddy:byte-buddy", version.ref = "bytebuddy" }
 cglib = "cglib:cglib:3.3.0"
+contextPropagation = "io.micrometer:context-propagation-api:1.0.0-SNAPSHOT"
```
Comment (violetagg) [inline gradle/libs.versions.toml:23]: Do we need this?

**Comment by simonbasle** [comment]
deferred to M3. we need to reach a broader consensus on the utility of this feature, from downstream stakeholder projects

**Comment by simonbasle** [comment]
this has never reached any traction, closing

**Comment by OlegDokuka** [comment]
now this PR should be relevant

**Comment by chemicL** [comment]
This PR has been closed and opened in the past. I still believe it is relevant, as the current context propagation with Scheduler hooks is based on `ThreadLocal` state instead of what is in the `Context`. However, unless there is a bug that surfaces, there might be no need for this at this moment in time.
The context-propagation theme was gathered at some point in #3468 for future reference.
Lack of this PR introduces sometimes unexpected, yet unspecified, behaviours, like the one documented in https://github.com/micrometer-metrics/context-propagation/issues/218.

We can try to recreate the change introduced in this PR in the future in case it is needed. Closing for now.

**Pull request closed without merging**